### DISCUSSION TITLE
UnusedFormalParameter should ignore public "constructor" as same as method.

### DISCUSSION THREAD

**Discussion by shu1saitohibm** [pmd/pmd#4769] [Q&A] [answered]
I'm having problem regarding "UnusedFormalParameter" rule.

I have an code with public constructor with unused parameter (like the following code), and UnusedFormalParameter rule detects this as error.
However, this is a "public" constructor that I already have provided to user, so I would not like to change this interface.
(since there are code that uses this constructor)

```java
public class Test {
   public Test(boolean testFlag) {
       ....
   }
}
```

The [document](https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#unusedformalparameter) mentions "checkAll" option and behaviour however it does not applied to "constructor" as I saw the latest [code.](https://github.com/pmd/pmd/blob/e4f2307b061ac440dd1cf1adc5f73a7b62fc99d2/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/UnusedFormalParameterRule.java#L32)

There is an if statement for "method", but the if statement is not present for "constructor".
```java
    @Override
    public Object visit(ASTConstructorDeclaration node, Object data) {
        check(node, data);
        return data;
    }

    @Override
    public Object visit(ASTMethodDeclaration node, Object data) {
        if (node.getVisibility() != Visibility.V_PRIVATE && !getProperty(CHECKALL_DESCRIPTOR)) {
            return data;
        }
      ...
```

Is this intended behaviour?

**Comment by adangel** [comment, marked as answer]
Thanks for bringing this up.

> Is this intended behaviour?

I think, that's an oversight in the rule implementation. Constructors should be considered here as well the same way, methods are handled.

I'll create an issue from this discussion. --> #4770

**Reply by shu1saitohibm** [reply]
@adangel Thank you for your comment, and opened the issue.
